src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6489,18 +6504,38 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
                 CheckDisallowedNullAssignment(rightState, leftAnnotations, right.Syntax.Location);
 
                 AdjustSetValue(left, declaredType, leftLValueType, ref rightState);
-                TrackNullableStateForAssignment(right, leftLValueType, MakeSlot(left), rightState, MakeSlot(right));
+                var leftSlot = MakeSlot(left);
+                TrackNullableStateForAssignment(right, leftLValueType, leftSlot, rightState, MakeSlot(right));
 
+                TypeWithAnnotations lvalueType;
+                TypeWithState resultState;
                 if (left is BoundDiscardExpression)
                 {
-                    var lvalueType = rightState.ToTypeWithAnnotations();
-                    SetResult(left, rightState, lvalueType, isLvalue: true);
-                    SetResult(node, rightState, lvalueType);
+                    lvalueType = rightState.ToTypeWithAnnotations();
+                    resultState = rightState;
                 }
                 else
                 {
-                    SetResult(node, TypeWithState.Create(leftLValueType.Type, rightState.State), leftLValueType);
+                    resultState = TypeWithState.Create(leftLValueType.Type, rightState.State);
+                    lvalueType = leftLValueType;
                 }
+
+
+                // We don't want to change the result of calling `GetTypeInfo` on the left side
+                // here if we can't observe that result in a later call to `GetTypeInfo` in a
+                // different context. For example:
+                //
+                //   var s = new string[1];
+                //   s[0] = null;
+                //   _ = s[0];
+                //
+                // GetTypeInfo on `s[0]` should return the same in both scenarios, and since we
+                // do not track nested array slots, it should be NotNull.
+                if (leftSlot > -1 || left is BoundDiscardExpression)
```

Comment:
Why do we want to do this for a `BoundDiscardExpression`? Not sure the value in setting the state for a discard. Do we still give IDE info for that?

Reply:
Yes, we still give the IDE info. You can mouse over a discard and see what type it was, which should be the the type/nullability that was "assigned".